### ISSUE TITLE
Execution pipeline

### ISSUE THREAD

**Issue by christopher-dG** [JuliaRegistries/TagBot#28]
There are a number of cases in which the webhook handler leaves a comment on a registry PR indicating something went wrong, with instructions to trigger a retry. In some of these cases (5xx from GitHub, `git` error), an automatic retry would be better.
We could take the same approach as the changelog generator, where instead of doing real work in the webhook handler, we just dump a message into SQS since it has better retry logic. In that case we just need to define which errors are retriable and which are not.

update: this is now basically a design refactor.  sqs is not really cost-effective and sns works essentially the same so the plan is to push some context message to sns topic(s) between steps. that way every step can have its own retry policy. Each step either has some side effect or 

1. webhook from github: validate and parse. keep computation minimal, since this can't retry.
2. notify: "starting", collect whatever github data will be needed (can retry, but should check for existing comments in case of retry)
3. git tag: skipped if a tag already exists, retries if necessary, absolute failure is ok too since an existing tag is not required
4. changelog generation: custom release notes are used if present,  otherwise use the generator. need the --future-release option since a release does not yet exist (i think? i don't know if a tag works without a release) failure = empty release notes + notification
5. github release: can retry, failure is actual failure
6. notify: success or failure, can retry

**Comment by christopher-dG** [comment]
Extending on this: it would be useful to break off certain bits of non-idempotent computation to avoid situations like this:

1. release created successfully
2. creating success comment gets a 5xx from GitHub
3. retry: release already exists so a success comment is not even attempted

**Comment by christopher-dG** [comment]
Some more random thoughts:

- [Step Functions](https://aws.amazon.com/step-functions/) might be a good way to model this kind of pipeline, but I'm not really sure how they work, there's a [plugin](https://github.com/horike37/serverless-step-functions) to support them
- Finding a way to make configuring "N retries, each with increasing memory and timeout" less verbose would be nice.

To illustrate, this:

```yaml
functions:
  foo:
    handler: foo.main
    scalingRetries:
      failureHandler: foo.failure
      pipeline:
        - timeout: 30
          memorySize: 1024
        - timeout: 45
          memorySize: 1536
        - timeout: 60
          memorySize: 2048
    events:
      - http:
          path: /foo
```

Could represent this:

```yaml
functions:
  foo_http:
    handler: foo.proxy
    events:
      - http:
          path: /foo
  foo1:
    handler: foo.main
    timeout: 30
    memorySize: 1024
    events:
      - sqs:
          arn: !GetAtt foo1_queue.Arn
  foo2:
    handler: foo.main
    timeout: 45
    memorySize: 1536
    events:
      - sqs:
          arn: !GetAtt foo2_queue.Arn
  foo3:
    handler: foo.main
    timeout: 60
    memorySize: 2048
    events:
      - sqs:
          arn: !GetAtt foo3_queue.Arn
  foo_failure:
    handler: foo.failure
    events:
      - sqs:
          arn: !GetAtt foodead_queue.Arn
resources:
  Resources:
    foo1_queue:
      Type: AWS::SQS::Queue
      Properties:
        VisibilityTimeout: 30
        RedrivePolicy:
          deadLetterTargetArn: !GetAtt foo2_queue.Arn
          maxReceiveCount: 1
    foo2_queue:
      Type: AWS::SQS::Queue
      Properties:
        VisibilityTimeout: 45
        RedrivePolicy:
          deadLetterTargetArn: !GetAtt foo3_queue.Arn
          maxReceiveCount: 1
    foo3_queue:
      Type: AWS::SQS::Queue
      Properties:
        VisibilityTimeout: 60
        RedrivePolicy:
          deadLetterTargetArn: !GetAtt foodead_queue.Arn
          maxReceiveCount: 1
    foodead_queue:
      Type: AWS::SQS::Queue
```

I think I could write a plugin for that, but writing nontrivial JS is not appealing to me. 
It would probably be easier to just operate on the file to convert from after to before.

**Comment by christopher-dG** [comment]
Alright, never mind all that YAML rewriting stuff, I just got a very well-timed email from AWS saying that I have used 85% of my free tier SQS capacity this month (in just 9 days). I did not realize this but SQS polls its queues constantly, so the number of requests made is not linear wrt the messages you put in. 

Keeping this free is high priority for me so I'll look into step functions. Even that has a fairly low free tier but I don't think it should be too hard to manage.
